src/python/dxpy/scripts/dx.py

```diff
@@ -2869,6 +2880,80 @@ def check_for_special_access(access_spec):
+    '''
+    Returns a dictionary that can be passed as a "systemRequirements" input to app-xxx/run, e.g.
+    {'fn': {'clusterSpec': {initialInstanceCount: 3, version: "2.4.0", ...}}}
+
+    Since full clusterSpec must be passed to the API server, we need to retrieve the cluster
+    spec defined in app doc's systemRequirements and overwrite the field initialInstanceCount
+    with the value the user passed to dx run for each entrypoint.
+
+    Instance count is currently the only clusterSpec's field the user is allowed to change
+    at runtime.
+
+    A few scenarios when requesting instance count for different entrypoints with dx run 
+    and the resulting merged systemRequirements (merged_cluster_spec):
+
+    Requested: {"*": 5}
+    App doc: {"main": "clusterSpec": {"initialInstanceCount": 7, bootstrapScript: "x.sh"},
+              "other": "clusterSpec": {"initialInstanceCount": 9, bootstrapScript: "y.sh"}}
+    Merged: {"main": "clusterSpec": {"initialInstanceCount": 5, bootstrapScript: "x.sh"},
+             "other": "clusterSpec": {"initialInstanceCount": 5, bootstrapScript: "y.sh"}}
+
+    Requested: {"main": 12}
+    App doc: {"main": "clusterSpec": {"initialInstanceCount": 7, bootstrapScript: "x.sh"},
+              "other": "clusterSpec": {"initialInstanceCount": 9, bootstrapScript: "y.sh"}}
+    Merged: {"main": "clusterSpec": {"initialInstanceCount": 12, bootstrapScript: "x.sh"},
+             "other": "clusterSpec": {"initialInstanceCount": 9, bootstrapScript: "y.sh"}}
+
+    Requested: {"main": 33}
+    App doc: {"*": "clusterSpec": {"initialInstanceCount": 2, bootstrapScript: "z.sh"}}
+    Merged: {"main": "clusterSpec": {"initialInstanceCount": 33, bootstrapScript: "z.sh"},
+             "*": "clusterSpec": {"initialInstanceCount": 2, bootstrapScript: "z.sh"}}
+
+    Requested: {"main": 22, "*": 11}
+    App doc: {"*": "clusterSpec": {"initialInstanceCount": 2, bootstrapScript: "t.sh"}}
+    Merged: {"main": "clusterSpec": {"initialInstanceCount": 22, bootstrapScript: "t.sh"},
+             "*": "clusterSpec": {"initialInstanceCount": 11, bootstrapScript: "t.sh"}}
+    '''
+
+    def replace_count_in_app_cluster_spec(merged_cluster_spec, app_sys_reqs, requested_count):
+        '''
+        Iterates over and updates all the app's entrypoints with the new, requested instance count.
+        Does nothing for entrypoints without clusterSpec.
+        '''
+
+        for app_entrypoint, reqs in app_sys_reqs.items():
+            if "clusterSpec" in reqs:
+                merged_cluster_spec[app_entrypoint] = {"clusterSpec": copy.deepcopy(reqs["clusterSpec"])}
+                merged_cluster_spec[app_entrypoint]["clusterSpec"]["initialInstanceCount"] = requested_count
+
+    merged_cluster_spec = {}
+    entrypoint_to_instance_count = instance_count_to_sys_reqs(instance_count)
+
+    for entrypoint, requested_count in entrypoint_to_instance_count.items():
+        if entrypoint == "*":
```

Comment:
seems like you can't guarantee that '*' is going to run first or after something else, right? In the case that it runs first, this works great - but if it comes up later, it will overwrite another entrypoint that may have set some other value.  Not totally sure if I'm correct on this, but just raising that it may be an issue.

Reply:
I have not reviewed the implementation details yet, but we could either process `"*"` up front, or skip it and process it at the end. Might simplify the implementation

Reply:
Thanks! I took it into account in the previous, slightly different implementation but then lost it when updating it. I now process if up front.